appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java

```diff
@@ -0,0 +1,141 @@
+// -*- mode: java; c-basic-offset: 2; -*-
+// Copyright 2020 MIT, All rights reserved
+// Released under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package com.google.appinventor.components.runtime;
+
+import com.google.appinventor.components.annotations.Options;
+import com.google.appinventor.components.annotations.SimpleEvent;
+import com.google.appinventor.components.annotations.SimpleFunction;
+import com.google.appinventor.components.annotations.SimpleProperty;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+import android.util.Log;
+
+public class OptionHelper {
+
+  public static final String TAG = "OptionHelper";
+
+  public static Map<String, Map<String, Method>> componentMethods =
+      new HashMap<String, Map<String, Method>>();
+    
+  /**
+   * Returns the OptionList version of the value if the function's return type has an @Options
+   * annotation notating that the value can be coerced to an OptionList.
+   */
+  public static <T> Object optionListFromValue(Component c, String func, T value) {
+    Method calledFunc = getMethod(c, func);
+    if (calledFunc == null) {
+      // Doesn't exist or not relevant.
+      return value;
+    }
+    Options annotation = calledFunc.getAnnotation(Options.class);
+    if (annotation == null) {
+      return value;
+    }
+    Class<?> optionListClass = annotation.value();
+    Object newValue = null;
+    try {
+      Method fromValue = optionListClass.getMethod("fromUnderlyingValue", value.getClass());
+      newValue = fromValue.invoke(optionListClass, value);
+    } finally {
+      if (newValue != null) {
+        return newValue;
+      }
+      return value;
+    }
+  }
+
+  /**
+   * Returns the args after any coercable args have been coerced to an OptionList. An arg is
+   * coercable if the function header has an @Options annotation associated with that arguement.
+   */
+  public static Object[] optionListsFromValues(Component c, String func, Object...args) {
+    if (args.length == 0) {
+      return args;
+    }
+    Method calledFunc = getMethod(c, func);
+    if (calledFunc == null) {
+      return args;
+    }
+    Annotation[][] paramAnnotations = calledFunc.getParameterAnnotations();
+    int i = 0;
+    for (Annotation[] annotations : paramAnnotations) {
+      for (Annotation annotation : annotations) {
+        if (annotation.annotationType() == Options.class) {
+          Options castAnnotation = (Options) annotation;
+          Class<?> optionListClass = castAnnotation.value();
+          try {
+            Method fromValue = optionListClass.getMethod("fromUnderlyingValue", args[i].getClass());
+            args[i] = fromValue.invoke(optionListClass, args[i]);
```

Comment:
I'm a bit confused about what you're saying. Shouldn't fromUnderlyingValue always return a valid OptionList in this case because this function is for coercing args of events? For example:

``` java
@SimpleEvent
public void MyTestEvent(@Options(SomeEnum.class) int val) {
  dispatchEvent(this, "MyTestEvent", val);
}
```

In the above case shouldn't val always be a valid underlying value of SomeEnum? If there aren't any bugs in the component that is.

I kind of feel like that's not what you're talking about though. Could you give me an example of the problem you're thinking of?

Reply:
There are a number of situations where we might have a closed set of known options but for whatever reason the option set itself is open world. One example is ScreenOrientation, where the set of options available at SDK 7 or whatever is smaller than it is today. Granted we enforce this to be closed even though newer versions of Android might add more ScreenOrientation options. The second scenario is where a value might be returned by an external-to-App Inventor component. For example, in the case of the Web component we have a set of known HttpStatus codes (e.g., HttpStatus.OK, HttpStatus.NotFound). However, various applications built on top of HTTP might return status codes not defined in the original spec. So if we were to introduce an options item for HttpStatus, which we might use with Web's GotText event with its responseCode parameter, there's a reasonable set of options that make sense/are common to expose as an option but the set is technically open. A server could return error code 599, for example, which might be specific to the server's software. I think it's still valuable for us to allow an option list for common values but not keep the set closed.

Reply:
Gotcha! I'll add logic to make sure that if we get a concrete value which isn't defined in the OptionList we return the concrete value.